File: python/geometer/_ipc_a0.py
"""Synchronous executable IPC A0 frame encoding and decoding."""

from __future__ import annotations

import struct
from dataclasses import dataclass
from enum import IntEnum
from typing import BinaryIO


HEADER_SIZE = 48
MAX_JSON_BYTES = 8 * 1024 * 1024
MAX_ATTACHMENT_COUNT = 16
MAX_ATTACHMENT_TEXT_BYTES = 128
MAX_ATTACHMENT_BYTES = 256 * 1024 * 1024
MAX_FRAME_BYTES = 512 * 1024 * 1024

_MAGIC = b"GMIPCA01"
_HEADER = struct.Struct("<8sHHHHQIIQII")
_ATTACHMENT_HEADER = struct.Struct("<HHIQ")


class FrameKind(IntEnum):
    HELLO = 1
    WELCOME = 2
    REQUEST = 3
    RESPONSE = 4
    CANCEL = 5
    CANCELLED = 6
    CANCEL_REJECTED = 7
    SHUTDOWN = 8
    SHUTDOWN_ACK = 9
    PROTOCOL_ERROR = 10


@dataclass(frozen=True, slots=True)
class Attachment:
    name: str
    media_type: str
    data: bytes


@dataclass(frozen=True, slots=True)
class Frame:
    kind: FrameKind
    request_id: int
    json: bytes
    attachments: tuple[Attachment, ...] = ()


class IpcFrameError(RuntimeError):
    """An executable IPC frame violates the governed A0 format."""


def encoded_size(frame: Frame) -> int:
    attachment_bytes = sum(_attachment_size(item) for item in frame.attachments)
    return HEADER_SIZE + len(frame.json) + attachment_bytes


def write_frame(stream: BinaryIO, frame: Frame) -> None:
    """Validate, write, and flush one complete IPC frame."""

    encoded_attachments = _validate_outgoing(frame)
    attachment_bytes = sum(
        _ATTACHMENT_HEADER.size + len(name) + len(media_type) + len(attachment.data)
        for attachment, name, media_type in encoded_attachments
    )
    header = _HEADER.pack(
        _MAGIC,
        HEADER_SIZE,
        0,
        int(frame.kind),
        0,
        frame.request_id,
        len(frame.json),
        len(frame.attachments),
        attachment_bytes,
        0,
        0,
    )
    _write_all(stream, header)
    _write_all(stream, frame.json)
    for attachment, name, media_type in encoded_attachments:
        _write_all(
            stream,
            _ATTACHMENT_HEADER.pack(len(name), len(media_type), 0, len(attachment.data)),
        )
        _write_all(stream, name)
        _write_all(stream, media_type)
        _write_all(stream, attachment.data)
    stream.flush()


def read_frame(
    stream: BinaryIO,
    *,
    max_json_bytes: int = MAX_JSON_BYTES,
    max_attachment_count: int = MAX_ATTACHMENT_COUNT,
    max_attachment_name_bytes: int = MAX_ATTACHMENT_TEXT_BYTES,
    max_attachment_media_type_bytes: int = MAX_ATTACHMENT_TEXT_BYTES,
    max_attachment_bytes: int = MAX_ATTACHMENT_BYTES,
    max_frame_bytes: int = MAX_FRAME_BYTES,
) -> Frame | None:
    """Read one complete IPC frame, returning ``None`` only at a clean EOF."""

    first = stream.read(1)
    if first == b"":
        return None
    if first is None:
        raise IpcFrameError("IPC stream returned no data without reaching EOF")
    header = first + _read_exact(stream, HEADER_SIZE - 1, "frame header")
    (
        magic,
        header_size,
        generation,
        raw_kind,
        flags,
        request_id,
        json_size,
        attachment_count,
        attachment_bytes,
        reserved0,
        reserved1,
    ) = _HEADER.unpack(header)
    if magic != _MAGIC:
        raise IpcFrameError("IPC frame magic is invalid")
    if header_size != HEADER_SIZE or generation != 0 or flags != 0 or reserved0 != 0 or reserved1 != 0:
        raise IpcFrameError("IPC frame header contains an unsupported or reserved value")
    try:
        kind = FrameKind(raw_kind)
    except ValueError as error:
        raise IpcFrameError("IPC frame kind is unknown") from error
    if json_size == 0 or json_size > max_json_bytes:
        raise IpcFrameError("IPC JSON section exceeds the A0 limit")
    if attachment_count > max_attachment_count or attachment_bytes > max_frame_bytes:
        raise IpcFrameError("IPC attachment section exceeds the A0 limit")
    complete_size = HEADER_SIZE + json_size + attachment_bytes
    if complete_size > max_frame_bytes:
        raise IpcFrameError("IPC complete frame exceeds the A0 limit")
    _validate_kind_and_request_id(kind, request_id)
    if kind not in {FrameKind.REQUEST, FrameKind.RESPONSE} and (attachment_count or attachment_bytes):
        raise IpcFrameError("IPC control frames cannot carry attachments")

    payload = _read_exact(stream, json_size + attachment_bytes, "frame payload")
    json_bytes = payload[:json_size]
    offset = json_size
    names: set[str] = set()
    attachments: list[Attachment] = []
    for _ in range(attachment_count):
        if len(payload) - offset < _ATTACHMENT_HEADER.size:
            raise IpcFrameError("IPC attachment header is truncated")
        name_size, media_type_size, attachment_flags, data_size = _ATTACHMENT_HEADER.unpack_from(payload, offset)
        offset += _ATTACHMENT_HEADER.size
        if (
            name_size > max_attachment_name_bytes
            or media_type_size > max_attachment_media_type_bytes
            or data_size > max_attachment_bytes
            or attachment_flags != 0
        ):
            raise IpcFrameError("IPC attachment header exceeds an A0 limit")
        section_size = name_size + media_type_size + data_size
        if section_size > len(payload) - offset:
            raise IpcFrameError("IPC attachment exceeds its declared section")
        name = _decode_text(payload[offset : offset + name_size], "attachment name")
        offset += name_size
        media_type = _decode_text(payload[offset : offset + media_type_size], "attachment media type")
        offset += media_type_size
        if not name or not media_type or name in names:
            raise IpcFrameError("IPC attachment metadata is empty or duplicated")
        names.add(name)
        data = bytes(payload[offset : offset + data_size])
        offset += data_size
        attachments.append(Attachment(name=name, media_type=media_type, data=data))
    if offset != len(payload) or offset - json_size != attachment_bytes:
        raise IpcFrameError("IPC attachment byte total does not match its sections")
    return Frame(kind=kind, request_id=request_id, json=json_bytes, attachments=tuple(attachments))


def _validate_outgoing(frame: Frame) -> list[tuple[Attachment, bytes, bytes]]:
    if not isinstance(frame.kind, FrameKind):
        raise IpcFrameError("IPC frame kind is invalid")
    if not isinstance(frame.request_id, int) or isinstance(frame.request_id, bool):
        raise IpcFrameError("IPC request id must be an unsigned integer")
    if not 0 <= frame.request_id <= (1 << 64) - 1:
        raise IpcFrameError("IPC request id exceeds uint64")
    if not isinstance(frame.json, bytes) or not frame.json or len(frame.json) > MAX_JSON_BYTES:
        raise IpcFrameError("IPC JSON section exceeds the A0 limit")
    if len(frame.attachments) > MAX_ATTACHMENT_COUNT:
        raise IpcFrameError("IPC attachment count exceeds the A0 limit")
    _validate_kind_and_request_id(frame.kind, frame.request_id)
    if frame.kind not in {FrameKind.REQUEST, FrameKind.RESPONSE} and frame.attachments:
        raise IpcFrameError("IPC control frames cannot carry attachments")

    names: set[str] = set()
    encoded: list[tuple[Attachment, bytes, bytes]] = []
    for attachment in frame.attachments:
        if not isinstance(attachment.data, bytes):
            raise IpcFrameError("IPC attachment data must be bytes")
        name = _encode_text(attachment.name, "attachment name")
        media_type = _encode_text(attachment.media_type, "attachment media type")
        if not name or not media_type or attachment.name in names:
            raise IpcFrameError("IPC attachment metadata is empty or duplicated")
        if (
            len(name) > MAX_ATTACHMENT_TEXT_BYTES
            or len(media_type) > MAX_ATTACHMENT_TEXT_BYTES
            or len(attachment.data) > MAX_ATTACHMENT_BYTES
        ):
            raise IpcFrameError("IPC attachment exceeds an A0 limit")
        names.add(attachment.name)
        encoded.append((attachment, name, media_type))
    if encoded_size(frame) > MAX_FRAME_BYTES:
        raise IpcFrameError("IPC complete frame exceeds the A0 limit")
    return encoded


def _validate_kind_and_request_id(kind: FrameKind, request_id: int) -> None:
    zero_id_kinds = {
        FrameKind.HELLO,
        FrameKind.WELCOME,
        FrameKind.SHUTDOWN,
        FrameKind.SHUTDOWN_ACK,
    }
    nonzero_id_kinds = {
        FrameKind.REQUEST,
        FrameKind.RESPONSE,
        FrameKind.CANCEL,
        FrameKind.CANCELLED,
        FrameKind.CANCEL_REJECTED,
    }
    if kind in zero_id_kinds and request_id != 0:
        raise IpcFrameError(f"IPC {kind.name.lower()} frame requires request id zero")
    if kind in nonzero_id_kinds and request_id == 0:
        raise IpcFrameError(f"IPC {kind.name.lower()} frame requires a nonzero request id")


def _attachment_size(attachment: Attachment) -> int:
    return (
        _ATTACHMENT_HEADER.size
        + len(attachment.name.encode())
        + len(attachment.media_type.encode())
        + len(attachment.data)
    )


def _read_exact(stream: BinaryIO, size: int, label: str) -> bytes:
    chunks = bytearray()
    while len(chunks) < size:
        chunk = stream.read(size - len(chunks))
        if not chunk:
            raise IpcFrameError(f"unexpected EOF within IPC {label}")
        chunks.extend(chunk)
    return bytes(chunks)


def _write_all(stream: BinaryIO, data: bytes) -> None:
    view = memoryview(data)
    while view:
        written = stream.write(view)
        if written is None or written <= 0:
            raise IpcFrameError("failed to write a complete IPC frame")
        view = view[written:]


def _encode_text(value: str, label: str) -> bytes:
    if not isinstance(value, str):
        raise IpcFrameError(f"IPC {label} must be text")
    try:
        return value.encode("utf-8", errors="strict")
    except UnicodeEncodeError as error:
        raise IpcFrameError(f"IPC {label} is not valid UTF-8") from error


def _decode_text(value: bytes, label: str) -> str:
    try:
        return value.decode("utf-8", errors="strict")
    except UnicodeDecodeError as error:
        raise IpcFrameError(f"IPC {label} is not valid UTF-8") from error

File: src/ts/geometer/ipc-a0.ts
const MAGIC = new Uint8Array([0x47, 0x4d, 0x49, 0x50, 0x43, 0x41, 0x30, 0x31]);
const HEADER_BYTES = 48;
const MAX_JSON_BYTES = 8 * 1024 * 1024;
const MAX_ATTACHMENT_COUNT = 16;
const MAX_ATTACHMENT_TEXT_BYTES = 128;
const MAX_ATTACHMENT_BYTES = 256 * 1024 * 1024;
const MAX_FRAME_BYTES = 512 * 1024 * 1024;
const textEncoder = new TextEncoder();
const textDecoder = new TextDecoder("utf-8", { fatal: true, ignoreBOM: true });

export const GEOMETER_IPC_A0_LIMITS = Object.freeze({
  attachmentBytes: MAX_ATTACHMENT_BYTES,
  attachmentCount: MAX_ATTACHMENT_COUNT,
  attachmentTextBytes: MAX_ATTACHMENT_TEXT_BYTES,
  frameBytes: MAX_FRAME_BYTES,
  jsonBytes: MAX_JSON_BYTES,
  pendingWriterBytes: MAX_FRAME_BYTES,
  queuedBytes: MAX_FRAME_BYTES,
  queuedRequests: 8,
  residentRequestBytes: MAX_FRAME_BYTES,
});

export type GeometerIpcFrameKind = 1 | 2 | 3 | 4 | 5 | 6 | 7 | 8 | 9 | 10;

export interface GeometerIpcAttachment {
  readonly name: string;
  readonly mediaType: string;
  readonly data: Uint8Array;
}

export interface GeometerIpcFrame {
  readonly kind: GeometerIpcFrameKind;
  readonly requestId: bigint;
  readonly json: string;
  readonly attachments: readonly GeometerIpcAttachment[];
}

export interface GeometerIpcFrameDecodeLimits {
  readonly attachmentBytes: number;
  readonly attachmentCount: number;
  readonly attachmentMediaTypeBytes: number;
  readonly attachmentNameBytes: number;
  readonly frameBytes: number;
  readonly jsonBytes: number;
}

export class GeometerIpcProtocolError extends Error {
  constructor(message: string) {
    super(message);
    this.name = "GeometerIpcProtocolError";
  }
}

export function validateIpcRequestOperationPair(envelope: IpcRequestA0): void {
  const declaration = operationDeclaration(envelope.operation);
  const projection = "requestProjection" in declaration ? declaration.requestProjection : undefined;
  const contract = validatePayloadProjection(
    envelope.request,
    declaration.runtimeDispatch,
    projection,
    declaration.runtimeDispatch === "logical_dto" ? declaration.requestContract : undefined,
  );
  if (contract !== declaration.requestContract) {
    throw new GeometerIpcProtocolError(
      `IPC request payload ${contract} does not match operation ${envelope.operation}.`,
    );
  }
}

export function validateIpcOutcomeOperationPair(outcome: OperationOutcomeA0): void {
  if (!outcome.ok) return;
  const declaration = operationDeclaration(outcome.operation);
  const projection = "resultProjection" in declaration ? declaration.resultProjection : undefined;
  const contract = validatePayloadProjection(
    outcome.result,
    declaration.runtimeDispatch,
    projection,
  );
  if (contract !== declaration.resultContract) {
    throw new GeometerIpcProtocolError(
      `IPC result payload ${contract} does not match operation ${outcome.operation}.`,
    );
  }
}

export function encodeGeometerIpcFrame(frame: GeometerIpcFrame): Uint8Array {
  validateKindAndRequestId(frame.kind, frame.requestId);
  const json = textEncoder.encode(frame.json);
  if (json.byteLength === 0 || json.byteLength > MAX_JSON_BYTES) {
    throw new GeometerIpcProtocolError("IPC JSON size is outside the A0 limit.");
  }
  validateAttachmentDirection(frame.kind, frame.attachments.length);
  if (frame.attachments.length > MAX_ATTACHMENT_COUNT) {
    throw new GeometerIpcProtocolError("IPC attachment count exceeds the A0 limit.");
  }

  const encodedAttachments: Array<{
    readonly name: Uint8Array;
    readonly mediaType: Uint8Array;
    readonly data: Uint8Array;
  }> = [];
  const names = new Set<string>();
  let attachmentBytes = 0;
  for (const attachment of frame.attachments) {
    if (names.has(attachment.name)) {
      throw new GeometerIpcProtocolError(`Duplicate IPC attachment ${attachment.name}.`);
    }
    names.add(attachment.name);
    const name = textEncoder.encode(attachment.name);
    const mediaType = textEncoder.encode(attachment.mediaType);
    if (
      name.byteLength === 0 ||
      name.byteLength > MAX_ATTACHMENT_TEXT_BYTES ||
      mediaType.byteLength === 0 ||
      mediaType.byteLength > MAX_ATTACHMENT_TEXT_BYTES
    ) {
      throw new GeometerIpcProtocolError("IPC attachment text is outside the A0 limit.");
    }
    if (attachment.data.byteLength > MAX_ATTACHMENT_BYTES) {
      throw new GeometerIpcProtocolError("IPC attachment data exceeds the A0 limit.");
    }
    attachmentBytes = checkedAdd(
      attachmentBytes,
      16 + name.byteLength + mediaType.byteLength + attachment.data.byteLength,
      "IPC attachment section",
    );
    encodedAttachments.push({ data: attachment.data, mediaType, name });
  }
  const frameBytes = checkedAdd(HEADER_BYTES + json.byteLength, attachmentBytes, "IPC frame");
  if (frameBytes > MAX_FRAME_BYTES) {
    throw new GeometerIpcProtocolError("IPC frame exceeds the A0 limit.");
  }

  const output = new Uint8Array(frameBytes);
  output.set(MAGIC, 0);
  const view = new DataView(output.buffer);
  view.setUint16(8, HEADER_BYTES, true);
  view.setUint16(10, 0, true);
  view.setUint16(12, frame.kind, true);
  view.setUint16(14, 0, true);
  view.setBigUint64(16, frame.requestId, true);
  view.setUint32(24, json.byteLength, true);
  view.setUint32(28, encodedAttachments.length, true);
  view.setBigUint64(32, BigInt(attachmentBytes), true);
  output.set(json, HEADER_BYTES);
  let offset = HEADER_BYTES + json.byteLength;
  for (const attachment of encodedAttachments) {
    view.setUint16(offset, attachment.name.byteLength, true);
    view.setUint16(offset + 2, attachment.mediaType.byteLength, true);
    view.setUint32(offset + 4, 0, true);
    view.setBigUint64(offset + 8, BigInt(attachment.data.byteLength), true);
    offset += 16;
    output.set(attachment.name, offset);
    offset += attachment.name.byteLength;
    output.set(attachment.mediaType, offset);
    offset += attachment.mediaType.byteLength;
    output.set(attachment.data, offset);
    offset += attachment.data.byteLength;
  }
  return output;
}

export class GeometerIpcFrameDecoder {
  private readonly pending = new ByteQueue();
  private limits: GeometerIpcFrameDecodeLimits = {
    attachmentBytes: MAX_ATTACHMENT_BYTES,
    attachmentCount: MAX_ATTACHMENT_COUNT,
    attachmentMediaTypeBytes: MAX_ATTACHMENT_TEXT_BYTES,
    attachmentNameBytes: MAX_ATTACHMENT_TEXT_BYTES,
    frameBytes: MAX_FRAME_BYTES,
    jsonBytes: MAX_JSON_BYTES,
  };

  /**
   * Transfers ownership of `chunk` to the decoder. The caller must not mutate or detach it after
   * this call. Decoded attachment views may retain the transferred storage without another copy.
   */
  push(chunk: Uint8Array): readonly GeometerIpcFrame[] {
    const frames: GeometerIpcFrame[] = [];
    let frame = this.pushOne(chunk);
    while (frame !== undefined) {
      frames.push(frame);
      frame = this.pushOne();
    }
    return frames;
  }

  /** Decodes at most one frame so negotiated limits can change between coalesced frames. */
  pushOne(chunk?: Uint8Array): GeometerIpcFrame | undefined {
    if (chunk !== undefined) this.pending.push(chunk);
    if (this.pending.byteLength < HEADER_BYTES) return undefined;
    const header = decodeFixedHeader(this.pending.peek(HEADER_BYTES), this.limits);
    if (this.pending.byteLength < header.frameBytes) return undefined;
    return decodeQueuedFrame(this.pending, header, this.limits);
  }

  setLimits(limits: GeometerIpcFrameDecodeLimits): void {
    const maxima: readonly (readonly [number, number])[] = [
      [limits.attachmentBytes, MAX_ATTACHMENT_BYTES],
      [limits.attachmentCount, MAX_ATTACHMENT_COUNT],
      [limits.attachmentMediaTypeBytes, MAX_ATTACHMENT_TEXT_BYTES],
      [limits.attachmentNameBytes, MAX_ATTACHMENT_TEXT_BYTES],
      [limits.frameBytes, MAX_FRAME_BYTES],
      [limits.jsonBytes, MAX_JSON_BYTES],
    ];
    if (
      maxima.some(([value, maximum]) => !Number.isInteger(value) || value <= 0 || value > maximum)
    ) {
      throw new GeometerIpcProtocolError("IPC decoder limit is outside the A0 maximum.");
    }
    this.limits = { ...limits };
  }

  finish(): void {
    if (this.pending.byteLength !== 0) {
      throw new GeometerIpcProtocolError("IPC stream ended within a frame.");
    }
  }
}

interface DecodedFixedHeader {
  readonly attachmentBytes: number;
  readonly attachmentCount: number;
  readonly frameBytes: number;
  readonly jsonBytes: number;
  readonly kind: GeometerIpcFrameKind;
  readonly requestId: bigint;
}

function decodeFixedHeader(
  bytes: Uint8Array,
  limits: GeometerIpcFrameDecodeLimits,
): DecodedFixedHeader {
  for (let index = 0; index < MAGIC.byteLength; index += 1) {
    if (bytes[index] !== MAGIC[index]) throw new GeometerIpcProtocolError("Invalid IPC magic.");
  }
  const view = new DataView(bytes.buffer, bytes.byteOffset, HEADER_BYTES);
  if (
    view.getUint16(8, true) !== HEADER_BYTES ||
    view.getUint16(10, true) !== 0 ||
    view.getUint16(14, true) !== 0 ||
    view.getUint32(40, true) !== 0 ||
    view.getUint32(44, true) !== 0
  ) {
    throw new GeometerIpcProtocolError("Unsupported IPC A0 header fields.");
  }
  const jsonBytes = view.getUint32(24, true);
  const attachmentCount = view.getUint32(28, true);
  const attachmentBytes = safeUint64(view.getBigUint64(32, true), "attachment section");
  const kindValue = view.getUint16(12, true);
  if (kindValue < 1 || kindValue > 10) {
    throw new GeometerIpcProtocolError("Unknown IPC frame kind.");
  }
  const kind = kindValue as GeometerIpcFrameKind;
  const requestId = view.getBigUint64(16, true);
  validateKindAndRequestId(kind, requestId);
  validateAttachmentDirection(kind, attachmentCount);
  if (jsonBytes === 0 || jsonBytes > limits.jsonBytes || attachmentCount > limits.attachmentCount) {
    throw new GeometerIpcProtocolError("IPC header declares an out-of-range payload.");
  }
  if ((attachmentCount === 0 && attachmentBytes !== 0) || attachmentBytes < attachmentCount * 18) {
    throw new GeometerIpcProtocolError("IPC attachment count and section size are inconsistent.");
  }
  const frameBytes = checkedAdd(HEADER_BYTES + jsonBytes, attachmentBytes, "IPC frame");
  if (frameBytes > limits.frameBytes) {
    throw new GeometerIpcProtocolError("IPC frame exceeds the A0 limit.");
  }
  return { attachmentBytes, attachmentCount, frameBytes, jsonBytes, kind, requestId };
}

function decodeQueuedFrame(
  queue: ByteQueue,
  header: DecodedFixedHeader,
  limits: GeometerIpcFrameDecodeLimits,
): GeometerIpcFrame {
  queue.read(HEADER_BYTES);
  let json: string;
  try {
    json = textDecoder.decode(queue.read(header.jsonBytes));
  } catch {
    throw new GeometerIpcProtocolError("IPC JSON is not valid UTF-8.");
  }
  const attachments: GeometerIpcAttachment[] = [];
  const names = new Set<string>();
  let attachmentBytesRead = 0;
  for (let index = 0; index < header.attachmentCount; index += 1) {
    const attachmentHeader = queue.read(16);
    attachmentBytesRead += 16;
    const view = new DataView(
      attachmentHeader.buffer,
      attachmentHeader.byteOffset,
      attachmentHeader.byteLength,
    );
    const nameBytes = view.getUint16(0, true);
    const mediaTypeBytes = view.getUint16(2, true);
    const flags = view.getUint32(4, true);
    const dataBytes = safeUint64(view.getBigUint64(8, true), "attachment data");
    if (
      flags !== 0 ||
      nameBytes === 0 ||
      nameBytes > limits.attachmentNameBytes ||
      mediaTypeBytes === 0 ||
      mediaTypeBytes > limits.attachmentMediaTypeBytes ||
      dataBytes > limits.attachmentBytes
    ) {
      throw new GeometerIpcProtocolError("IPC attachment header is invalid.");
    }
    const payloadBytes = checkedAdd(nameBytes + mediaTypeBytes, dataBytes, "IPC attachment");
    attachmentBytesRead = checkedAdd(attachmentBytesRead, payloadBytes, "IPC attachment section");
    if (attachmentBytesRead > header.attachmentBytes) {
      throw new GeometerIpcProtocolError("IPC attachment exceeds its declared section.");
    }
    let name: string;
    let mediaType: string;
    try {
      name = textDecoder.decode(queue.read(nameBytes));
      mediaType = textDecoder.decode(queue.read(mediaTypeBytes));
    } catch {
      throw new GeometerIpcProtocolError("IPC attachment text is not valid UTF-8.");
    }
    if (names.has(name)) throw new GeometerIpcProtocolError(`Duplicate IPC attachment ${name}.`);
    names.add(name);
    const data = queue.read(dataBytes);
    attachments.push({ data, mediaType, name });
  }
  if (attachmentBytesRead !== header.attachmentBytes) {
    throw new GeometerIpcProtocolError("IPC attachment section length does not match its header.");
  }
  return { attachments, json, kind: header.kind, requestId: header.requestId };
}

class ByteQueue {
  private readonly chunks: Uint8Array[] = [];
  private headIndex = 0;
  private headOffset = 0;
  byteLength = 0;

  push(chunk: Uint8Array): void {
    if (chunk.byteLength === 0) return;
    this.chunks.push(chunk);
    this.byteLength = checkedAdd(this.byteLength, chunk.byteLength, "IPC stream buffer");
  }

  peek(length: number): Uint8Array {
    return this.collect(length, false);
  }

  read(length: number): Uint8Array {
    return this.collect(length, true);
  }

  private collect(length: number, consume: boolean): Uint8Array {
    if (!Number.isSafeInteger(length) || length < 0 || length > this.byteLength) {
      throw new GeometerIpcProtocolError("IPC byte queue underflow.");
    }
    if (length === 0) return new Uint8Array(0);
    const first = this.chunks[this.headIndex];
    if (first === undefined) throw new GeometerIpcProtocolError("IPC byte queue underflow.");
    const available = first.byteLength - this.headOffset;
    if (available >= length) {
      const output = first.subarray(this.headOffset, this.headOffset + length);
      if (consume) this.discard(length);
      return output;
    }

    const output = new Uint8Array(length);
    let copied = 0;
    let chunkIndex = this.headIndex;
    let offset = this.headOffset;
    while (copied < length) {
      const chunk = this.chunks[chunkIndex];
      if (chunk === undefined) throw new GeometerIpcProtocolError("IPC byte queue underflow.");
      const count = Math.min(chunk.byteLength - offset, length - copied);
      output.set(chunk.subarray(offset, offset + count), copied);
      copied += count;
      chunkIndex += 1;
      offset = 0;
    }
    if (consume) this.discard(length);
    return output;
  }

  private discard(length: number): void {
    let remaining = length;
    while (remaining !== 0) {
      const first = this.chunks[this.headIndex];
      if (first === undefined) throw new GeometerIpcProtocolError("IPC byte queue underflow.");
      const available = first.byteLength - this.headOffset;
      if (remaining < available) {
        this.headOffset += remaining;
        remaining = 0;
      } else {
        remaining -= available;
        this.headIndex += 1;
        this.headOffset = 0;
      }
    }
    this.byteLength -= length;
    if (this.headIndex === this.chunks.length) {
      this.chunks.length = 0;
      this.headIndex = 0;
    } else if (this.headIndex >= 1024 && this.headIndex * 2 >= this.chunks.length) {
      this.chunks.splice(0, this.headIndex);
      this.headIndex = 0;
    }
  }
}

function validateKindAndRequestId(kind: GeometerIpcFrameKind, requestId: bigint): void {
  if (!Number.isInteger(kind) || kind < 1 || kind > 10) {
    throw new GeometerIpcProtocolError("Unknown IPC frame kind.");
  }
  if (requestId < 0n || requestId > 0xffff_ffff_ffff_ffffn) {
    throw new GeometerIpcProtocolError("IPC request id is outside uint64.");
  }
  const requiresZero = kind === 1 || kind === 2 || kind === 8 || kind === 9;
  const requiresNonzero = kind >= 3 && kind <= 7;
  if ((requiresZero && requestId !== 0n) || (requiresNonzero && requestId === 0n)) {
    throw new GeometerIpcProtocolError("IPC request id does not match the frame kind.");
  }
}

function validateAttachmentDirection(kind: GeometerIpcFrameKind, count: number): void {
  if (count !== 0 && kind !== 3 && kind !== 4) {
    throw new GeometerIpcProtocolError("IPC control frames cannot carry attachments.");
  }
}

function safeUint64(value: bigint, field: string): number {
  if (value > BigInt(Number.MAX_SAFE_INTEGER)) {
    throw new GeometerIpcProtocolError(`IPC ${field} exceeds safe integer range.`);
  }
  return Number(value);
}

function checkedAdd(left: number, right: number, field: string): number {
  const value = left + right;
  if (!Number.isSafeInteger(value)) {
    throw new GeometerIpcProtocolError(`${field} size overflow.`);
  }
  return value;
}

function operationDeclaration(operation: string) {
  if (!Object.hasOwn(operationCatalog, operation)) {
    throw new GeometerIpcProtocolError(`Unknown IPC operation ${operation}.`);
  }
  return operationCatalog[operation as OperationIdentity];
}

function validatePayloadProjection(
  value: object,
  runtimeDispatch: "logical_dto" | "packed_attachment",
  expectedProjection?: {
    readonly attachment_name: string;
    readonly format: string;
    readonly kind: "packed_attachment";
  },
  fallback?: string,
): string {
  const hasPacket = "packet" in value;
  if (runtimeDispatch === "logical_dto") {
    if (hasPacket) {
      throw new GeometerIpcProtocolError(
        "A logical DTO operation cannot carry a packed attachment projection.",
      );
    }
  } else {
    if (!hasPacket || expectedProjection === undefined) {
      throw new GeometerIpcProtocolError("A packed operation requires its declared projection.");
    }
    const packet = value.packet;
    if (
      typeof packet !== "object" ||
      packet === null ||
      !("attachment" in packet) ||
      packet.attachment !== expectedProjection.attachment_name ||
      !("format" in packet) ||
      packet.format !== expectedProjection.format
    ) {
      throw new GeometerIpcProtocolError(
        "Packed attachment projection metadata differs from the operation declaration.",
      );
    }
  }
  if ("schema" in value && typeof value.schema === "string") return value.schema;
  if (fallback !== undefined) return fallback;
  throw new GeometerIpcProtocolError("IPC payload has no contract schema identity.");
}

import type { IpcRequestA0, OperationOutcomeA0 } from "./generated/contracts.js";
import { type OperationIdentity, operationCatalog } from "./generated/operations.js";
